volatility3/framework/plugins/windows/mbrscan.py

```diff
@@ -0,0 +1,109 @@
+        ]
+
+    @classmethod
+    def get_hash(cls, data:bytes) -> str:
+        return hashlib.md5(data).hexdigest()
+
+    def _generator(self):
+        kernel = self.context.modules[self.config['kernel']]
+        physical_layer_name = self.context.layers[kernel.layer_name].config.get('memory_layer', None)
+        
+        # Decide of Memory Dump Architecture
+        layer = self.context.layers[physical_layer_name]
+        architecture = "intel" if not symbols.symbol_table_is_64bit(self.context, kernel.symbol_table_name) else "intel64"
+
+        # Read in the Symbol File
+        symbol_table = intermed.IntermediateSymbolTable.create(context = self.context,
+                                                               config_path = self.config_path,
+                                                               sub_path = "windows",
+                                                               filename = "mbr",
+                                                               class_types = {
+                                                                'PARTITION_TABLE': mbr.PARTITION_TABLE,
+                                                                'PARTITION_ENTRY': mbr.PARTITION_ENTRY
+                                                               })
+
+        partition_table_object = symbol_table + constants.BANG + "PARTITION_TABLE"
+        
+        # Define Signature and Data Length
+        mbr_signature = b"\x55\xAA"
+        mbr_length = 0x200
+        bootcode_length = 0x1B8
+
+        # Scan the Layer for Raw Master Boot Record (MBR) and parse the fields
+        for offset, _value in layer.scan(context = self.context, scanner = scanners.MultiStringScanner(patterns = [mbr_signature])):
+            try:
+                mbr_start_offset = offset - (mbr_length - len(mbr_signature))
+                partition_table = self.context.object(partition_table_object, offset = mbr_start_offset, layer_name = layer.name)
+
+                # Extract only BootCode
+                full_mbr = layer.read(mbr_start_offset, mbr_length, pad = True)
+                bootcode = full_mbr[:bootcode_length]
+                
+                if bootcode:
+                    all_zeros = bootcode.count(b"\x00") == len(bootcode)
+
+                if not all_zeros:
+                    partition_entries = [ 
+                                            partition_table.FirstEntry,
+                                            partition_table.SecondEntry,
+                                            partition_table.ThirdEntry,
+                                            partition_table.FourthEntry
+                                        ]
+                    partition_info = "\n"
+
+                    for index, partition_entry_object in enumerate(partition_entries):
+                        partition_entry_object.set_index(index)
+                        partition_info += str(partition_entry_object)
```

Comment:
This looks like it accumulates a lot of information into a string.  This requires any libraries or programs that consume the output of this plugin to reparse all that data, we're not writing the output for human consumption, we're writing it to be consumed by anyone (programs or humans).

As such, please could you have have additional rows, something along the lines of:

```
ptable1 tabledata1 partition1 partition_attribute1
ptable1 tabledata1 partition2 partition_attribute2
ptable1 tabledata1 partition3 partition_attribute3
ptable2 tabledata2 partition1 partition_attribute1
```
This will allow humans to read the information if needed, but also allows others to build programs that can digest and use the information in its original form?  It's fine to do things like looking up types and converting them to strings, just leave each item of data as its own column, rather than jamming everything into a string with specific formatting.  We've no idea how a web interface may decide to render that line, so much better to keep the data as small and specific as possible.

Reply:
I fully sympathize with what you said, and I tried to modify it based on this.
The current code was organized in a simple string table, so it was easy to provide a lot of information,
but it was suggested that all the information about the four partitions be provided in separate rows because it takes resources to reuse programs and people.

So, as a result of experimenting with this, there was some inconvenience for the CLI user.
(4 Partition Entries * 10 Info) 40 informations were printed on the CLI as information about the partition table, making it difficult to distinguish information. (I think this is the limitation of the CLI that the more information it contains, the less communicative it is.)

Here are the results of considering the appropriate way to satisfy all of the various use cases. In terms of memory forensics and digital forensics, the need for MBR partition table information is more important as to which partitions exist on the system, whether they are bootable, and how large they are,
so I don't think it's necessary to provide all the information.

Therefore, why don't you yield only the three information (**BootFlag, PartitionType, SizeInSector**) and leave the code for the rest of the information to be called separately?

Reply:
Sure, I think that's ok.  You could hide the rest of the information behind a flag like `--full` or something?  Also be aware that while the CLI output is limited, there are options such as `-r csv` or `-r pretty` that might improve the output.

Reply:
Thanks for the idea, I think the option you mentioned will help us overcome the limitations of the CLI and provide good data to everyone. I will add related options.

Reply:
When I converted it to CSV format according to the feedback, I got a much better result!
Rendering in CSV format seems to be very good for providing users with a large amount of data in addition to providing it in table format.

**Command**

```
python3 vol.py -f test.vmem -r csv windows.mbrscan --full > test.csv
```

**Screenshot**
<img width="990" alt="스크린샷 2022-03-17 오전 2 31 05" src="https://user-images.githubusercontent.com/16293464/158652810-460dd9c2-4973-44f4-bb0c-1a2cca050e42.png">